packages/react-components/react-jsx-runtime/src/createElement.ts

```diff
@@ -0,0 +1,49 @@
+import * as React from 'react';
+import { SlotRenderFunction, UnknownSlotProps, SLOT_RENDER_FUNCTION_SYMBOL } from '@fluentui/react-utilities';
+
+type WithMetadata<Props extends {}> = Props & {
+  [SLOT_RENDER_FUNCTION_SYMBOL]: SlotRenderFunction<Props>;
+};
+
+export function createElement<P extends {}>(
+  type: React.ElementType<P>,
+  props?: P | null,
+  ...children: React.ReactNode[]
+): React.ReactElement<P> | null {
+  if (!isSlotComponent(props)) {
+    return React.createElement(type, props, ...children);
+  }
+
+  const result = normalizeRenderFunction(props, children);
+  return React.createElement(
+    React.Fragment,
+    {},
+    result.renderFunction(type, { ...result.props, children: result.children }),
+  ) as React.ReactElement<P>;
+}
+
+function normalizeRenderFunction<Props extends UnknownSlotProps>(
+  propsWithMetadata: WithMetadata<Props>,
+  overrideChildren?: React.ReactNode[],
+): {
+  props: Props;
+  children: React.ReactNode;
+  renderFunction: SlotRenderFunction<Props>;
+} {
+  const { [SLOT_RENDER_FUNCTION_SYMBOL]: renderFunction, children: externalChildren, ...props } = propsWithMetadata;
+
+  const children: React.ReactNode =
+    Array.isArray(overrideChildren) && overrideChildren.length > 0
+      ? React.createElement(React.Fragment, {}, ...overrideChildren)
+      : externalChildren;
+
+  return {
+    children,
+    renderFunction,
+    props: props as UnknownSlotProps as Props,
+  };
+}
```

Comment:
It seems like these functions could be simplified quite a bit, to reduce the number of Fragments and object spreads, and to be easier to understand:
```tsx
export function createElement<P extends {}>(
  type: React.ElementType<P>,
  props?: P | null,
  ...children: React.ReactNode[]
): React.ReactElement<P> | null {
  if (!isSlotComponent(props)) {
    return React.createElement(type, props, children);
  }

  const { [SLOT_RENDER_FUNCTION_SYMBOL]: renderFunction, ...renderProps } = props;

  if (Array.isArray(children) && children.length > 0) {
    renderProps.children = children;
  }

  return React.createElement(React.Fragment, {}, renderFunction(type, renderProps));
}
```

Reply:
`renderProps.children` **cannot** be an array! it has to be converted to a fragment. otherwise we'll be requiring for  the render function to properly provide keys for each children element.

in this simplification, `renderProps.children` might end up being an array.

Our tests are even catching this problem:

![image](https://user-images.githubusercontent.com/5483269/231985519-eb85321d-5dbc-4615-9d47-6b3d7451b038.png)


Nonetheless I can see clear simplifications with this method. I'll see to follow up on this with another PR